genie-proto/src/main/proto/genie.proto

```diff
@@ -52,8 +52,11 @@ message AgentRegistrationResponse {
 // Job Service
 service JobService {
     rpc reserveJobId (ReserveJobIdRequest) returns (ReserveJobIdResponse);
-    rpc resolveJobSpecification (ResolveJobSpecificationRequest) returns (JobSpecificationResponse);
-    rpc getJobSpecification (GetJobSpecificationRequest) returns (JobSpecificationResponse);
+    rpc resolveJobSpecification (JobSpecificationRequest) returns (JobSpecificationResponse);
+    rpc getJobSpecification (JobSpecificationRequest) returns (JobSpecificationResponse);
+    rpc resolveJobSpecificationDryRun (DryRunJobSpecificationRequest) returns (JobSpecificationResponse);
```

Comment:
I'm not sure why you need all 3:

```
rpc resolveJobSpecification (JobSpecificationRequest) returns (JobSpecificationResponse);
rpc getJobSpecification (JobSpecificationRequest) returns (JobSpecificationResponse);
rpc resolveJobSpecificationDryRun (DryRunJobSpecificationRequest) returns (JobSpecificationResponse);
```

The first one in particular.
In case of agent job, we send the "Job request" with the ID reservatuin, and the agent later can use `getJobSpecification`.
In case of job via API, this is not even invoked, the agent has an ID and can again just use `getJobSpecification`.

Could we get rid of `resolveJobSpecificationDryRun` and make `resolveJobSpecification` a side-effect-free API that just resolves without persisting?

Reply:
There are three because of the three existing cases (in API order):
1. Agent launches and reserves job id. Record exists in database but job specification hasn't been calculated yet. Agent calls this API to kick off the job specification resolution based on the values saved in the database. Specification details saved in database and returned.
2. Agent launched after API call. In this case server already did job specification resolution before agent was lost so we're just calling `get` which returns the values that were saved in the database previously and return them without resolving again
3. Dry run which requires sending a bunch of information that wouldn't be in the database cause we never reserved id or ran through the API so we have to send it all and only run the resolution side effect free

I thought we'd agreed on this but if we're still not clear we can discuss further

Reply:
I guess we didn't agree on a small detail: in the first case (agent claiming an ID) you were thinking the "job request" is just parked in the DB until we explicitly call resolve.
I was thinking we could resolve it then and there and avoid "burning" the ID if the request is not resolving. If it does resolve, we store request and spec, and the agent just need to do a `get` later, thus making the first of these three unnecessary.
You were instead thinking of just storing the spec and don't attempt to resolve until explicitly asked.
Either works... 
I just find easier to keep in mind that the agent always just need to do a `get`.
And `resolve` is just an API to resolve without side-effects.